### PR TITLE
Fix issue 1681: MsBuildUtility should ignore invalid msbuild versions…

### PR DESCRIPTION
**Pull request by feiling**
… instead of throwing.

TryParse() is used to process toolset.ToolsVersions.

@emgarten @yishaigalatzer @MeniZalzman @deepakaravindr

### CODE REVIEW

File: src/NuGet.Clients/NuGet.CommandLine/MsBuildUtility.cs [hunk 1]
```diff
@@ -258,16 +261,32 @@ private static Version GetMSBuildVersionInPath()
...
+                    toolset => SafeParseVersion(toolset.ToolsVersion)).ToList();
             }
 
-            if (string.IsNullOrEmpty(version))
+            return GetMsbuildDirectoryInternal(userVersion, console, installedToolsets);
+        }
+
+        // This method is called by GetMsbuildDirectory(). This method is not intended to be called directly.
+        // It's marked public so that it can be called by unit tests.
+        public static string GetMsbuildDirectoryInternal(string userVersion, IConsole console, IEnumerable<Toolset> installedToolsets)
```
Comment (yishaigalatzer) [inline src/NuGet.Clients/NuGet.CommandLine/MsBuildUtility.cs]: long line

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/MSBuildUtilityTest.cs [hunk 2]
```diff
@@ -135,5 +135,153 @@ private void HighestVersionSelectedIfNoVersionMatch()
...
+
+                // Assert
+                Assert.Equal(directory, toolsetFoo4.ToolsPath);
+            }
+        }
+
+        // Tests that, when userVersion cannot be parsed into Version, string comparison is used 
+        // to match toolset and ".0" is not appended during matching.
+        [Fact]
+        public void TestVersionMatchByStringExact()
```
Comment (yishaigalatzer) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/MSBuildUtilityTest.cs]: add a theory test here, and pass the string with different cases, and additional space before/after

**Comment by yishaigalatzer** [comment]
:shipit: when manual tests are run

File: src/NuGet.Clients/NuGet.CommandLine/MsBuildUtility.cs [hunk 3]
```diff
@@ -296,35 +315,47 @@ public static string GetMsbuildDirectory(string version, IConsole console)
...
-                        var toolsVersion = new Version(toolset.ToolsVersion);
+                        return true;
+                    }
+
+                    if (hasNumericVersion)
+                    {
+                        // then match by Major & Minor version numbers.
+                        var toolsVersion = SafeParseVersion(toolset.ToolsVersion);
+
                         return (toolsVersion.Major == ver.Major &&
```
Comment (emgarten) [inline src/NuGet.Clients/NuGet.CommandLine/MsBuildUtility.cs]: If the user specifies -MSBuildVersion 0  and has multiple bad strings in the registry, will nuget try to use the first bad one it hits? SafeParseVersion will return 0.0, and that will match the user specified one.

Comment (feiling) [inline src/NuGet.Clients/NuGet.CommandLine/MsBuildUtility.cs]: You're right. Fixed so major & minor numbers match is used only when both the userVersion and toolset.ToolsVersion are valid Versions.

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/MSBuildUtilityTest.cs (hunk 2 above)
Comment (feiling) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/MSBuildUtilityTest.cs]: Changed tests to use Theory.

**Comment by feiling** [comment]
Code changed per comments. Also manually verified that the fix worked.

**Comment by emgarten** [comment]
:shipit:

**Comment by feiling** [comment]
Fix rebased and merged.

**Pull request closed without merging**